pypsrp/wsman.py

```diff
@@ -126,6 +126,7 @@ def __init__(self, server, max_envelope_size=153600, operation_timeout=20,
                  path="wsman", auth="negotiate", cert_validation=True,
                  connection_timeout=30, encryption='auto', proxy=None,
                  no_proxy=False, locale='en-US', data_locale=None,
+                 reconnection_retries=4, reconnection_backoff=2.0,
```

Comment:
What's your reasoning for making this the default? I'm leaning towards keeping the existing behaviour rather than defaulting to retry on newer versions.

Reply:
I think reliability is more important than keeping the existing behavior, especially if the new behavior is identical to the existing behavior when everything works fine. Only when things don't work fine immediately (Connection Refused, Service Unavailable, Too Early, Too Many Requests) we retry, which is exactly what we should be doing (knowing what we know now).

The reason for choosing 4 retries with a backoff of 2 seconds means we will retry 4 times no longer than 30 seconds. Which is exactly the same as we have specified for a connection timeout. So I think this acceptable, if you accept that an unresponsive system will also take 30 seconds by default before it fails.

Not retrying in the above cases is the wrong behavior IMO.

Reply:
The documentation is not clear on this, but normally the connection timeout is per connection (not the total time) this should be apparent when you have slow DNS lookups or DNS problems, in this case the DNS latency is added to the connection timeout as well.

If you use retry=4, each connection has the same timeout. For the successful case there's no difference in behavior, for the case where the connection was established (but faulty) there's also no difference in behavior. But for the case where the server does not respond, and retries are attempted (not sure if it retries in this case, I would assume not) the connection timeout could be 4x as long. But it's easy to test this behavior in fact.

Reply:
Let's try it out, I'm wondering whether this will affect the logic in win_reboot and we should override some values in there.

Reply:
Yeah, we may need to look at win_reboot, although the default 4 retries and backoff value of 2 seconds works fine in our tests for win_reboot too. But it's probably not really needed.

Reply:
So in one of my tests connecting with WinRM to a non-existing port, it timed out after 30 seconds. Which means that it doesn't timeout for every retry, the timeout appears to cover the total time (including retries). Not sure whether this is because of urllib3, requests or a lower-level mechanism in this case.